File: unity/com.dronecv.flight/Runtime/Protocol/Messages.cs
// dronecv wire protocol v1.0 message headers.
// Mirror of src/dronecv/protocol/messages.py — keep the two in sync and bump
// ProtocolVersion together. Field names must match the JSON exactly
// (snake_case), hence the JsonProperty annotations.

using System.Collections.Generic;
using Newtonsoft.Json;

namespace DroneCV.Flight.Protocol
{
    public static class ProtocolConstants
    {
        public const string ProtocolVersion = "1.0";
        public const int MaxHeaderLen = 1 << 20;
        public const long MaxBlobLen = 1L << 28;
    }

    public class BlobSpec
    {
        [JsonProperty("name")] public string Name;
        [JsonProperty("dtype")] public string Dtype = "bytes";        // uint8 | float32 | bytes
        [JsonProperty("shape")] public int[] Shape;                    // null for encoded streams
        [JsonProperty("encoding")] public string Encoding = "raw";     // raw | jpeg | png
        [JsonProperty("byte_len")] public long ByteLen;
    }

    public class Header
    {
        [JsonProperty("type")] public string Type;
        [JsonProperty("version")] public string Version = ProtocolConstants.ProtocolVersion;
        [JsonProperty("msg_id")] public long MsgId;
        [JsonProperty("sim_time")] public double SimTime;
        [JsonProperty("blobs")] public List<BlobSpec> Blobs = new List<BlobSpec>();
    }

    public class Hello : Header
    {
        public Hello() { Type = "hello"; }
        [JsonProperty("role")] public string Role = "localizer";
    }

    public class CameraInfo
    {
        [JsonProperty("width")] public int Width;
        [JsonProperty("height")] public int Height;
        [JsonProperty("fov_deg")] public float FovDeg;
        [JsonProperty("tilt_deg")] public float TiltDeg;
    }

    public class HelloAck : Header
    {
        public HelloAck() { Type = "hello_ack"; }
        [JsonProperty("sim_kind")] public string SimKind = "unity";
        [JsonProperty("env_id")] public string EnvId;
        [JsonProperty("capabilities")] public List<string> Capabilities = new List<string>();
        [JsonProperty("geo_meta", NullValueHandling = NullValueHandling.Include)]
        public Dictionary<string, double> GeoMeta;
        [JsonProperty("camera")] public CameraInfo Camera;
    }

    public class EnvInfoRequest : Header
    {
        public EnvInfoRequest() { Type = "env_info_request"; }
    }

    public class EnvInfo : Header
    {
        public EnvInfo() { Type = "env_info"; }
        [JsonProperty("bounds_min_sim")] public double[] BoundsMinSim;
        [JsonProperty("bounds_max_sim")] public double[] BoundsMaxSim;
        [JsonProperty("ground_alt_min_m")] public double GroundAltMinM;
        [JsonProperty("ground_alt_max_m")] public double GroundAltMaxM;
    }

    public class Subscribe : Header
    {
        public Subscribe() { Type = "subscribe"; }
        [JsonProperty("channels")] public List<string> Channels = new List<string>();
    }

    public class SubscribeAck : Header
    {
        public SubscribeAck() { Type = "subscribe_ack"; }
        [JsonProperty("channels")] public List<string> Channels = new List<string>();
    }

    public class CaptureRequest : Header
    {
        public CaptureRequest() { Type = "capture_request"; }
        [JsonProperty("pos_sim")] public double[] PosSim;
        [JsonProperty("yaw_deg")] public double YawDeg;
        [JsonProperty("pitch_deg")] public double PitchDeg;
        [JsonProperty("want")] public List<string> Want = new List<string> { "rgb" };
    }

    public class CaptureResult : Header
    {
        public CaptureResult() { Type = "capture_result"; }
        [JsonProperty("pos_sim")] public double[] PosSim;
        [JsonProperty("quat_sim")] public double[] QuatSim;
        [JsonProperty("yaw_deg")] public double YawDeg;
        [JsonProperty("pitch_deg")] public double PitchDeg;
        [JsonProperty("utc")] public string Utc;
        [JsonProperty("sun_azimuth_deg", NullValueHandling = NullValueHandling.Include)]
        public double? SunAzimuthDeg;
        [JsonProperty("sun_elevation_deg", NullValueHandling = NullValueHandling.Include)]
        public double? SunElevationDeg;
        [JsonProperty("moon_azimuth_deg", NullValueHandling = NullValueHandling.Include)]
        public double? MoonAzimuthDeg;
        [JsonProperty("moon_elevation_deg", NullValueHandling = NullValueHandling.Include)]
        public double? MoonElevationDeg;
    }

    public class SensorFrame : Header
    {
        public SensorFrame() { Type = "sensor_frame"; }
        [JsonProperty("frame_id")] public long FrameId;
        [JsonProperty("utc")] public string Utc;
        [JsonProperty("lidar_range_m", NullValueHandling = NullValueHandling.Include)]
        public double? LidarRangeM;
        // Deliberately NO pose fields: the localizer flies blind.
    }

    public class Command : Header
    {
        public Command() { Type = "command"; }
        [JsonProperty("vel_sim", NullValueHandling = NullValueHandling.Include)]
        public double[] VelSim;
        [JsonProperty("yaw_rate_dps", NullValueHandling = NullValueHandling.Include)]
        public double? YawRateDps;
    }

    public class TruthState : Header
    {
        public TruthState() { Type = "truth_state"; }
        [JsonProperty("frame_id")] public long FrameId;
        [JsonProperty("pos_sim")] public double[] PosSim;
        [JsonProperty("quat_sim")] public double[] QuatSim;
        [JsonProperty("vel_sim")] public double[] VelSim;
        [JsonProperty("collided")] public bool Collided;
    }

    public class Reset : Header
    {
        public Reset() { Type = "reset"; }
        [JsonProperty("seed")] public int Seed;
        [JsonProperty("start_pos_sim", NullValueHandling = NullValueHandling.Include)]
        public double[] StartPosSim;
        [JsonProperty("start_yaw_deg")] public double StartYawDeg;
        [JsonProperty("utc", NullValueHandling = NullValueHandling.Include)]
        public string Utc;
    }

    public class ResetDone : Header
    {
        public ResetDone() { Type = "reset_done"; }
        [JsonProperty("pos_sim")] public double[] PosSim;
    }

    public class Teleport : Header
    {
        public Teleport() { Type = "teleport"; }
        [JsonProperty("pos_sim")] public double[] PosSim;
        [JsonProperty("yaw_deg")] public double YawDeg;
    }

    public class ErrorMsg : Header
    {
        public ErrorMsg() { Type = "error"; }
        [JsonProperty("message")] public string Message;
    }

    public class Ping : Header { public Ping() { Type = "ping"; } }
    public class Pong : Header { public Pong() { Type = "pong"; } }
}

File: src/dronecv/protocol/messages.py
"""Versioned wire messages shared by the Unity bridge and the headless sim.

Every message is a JSON header, optionally followed by binary blobs described
by `blobs`. The Python side cannot tell Unity and the headless simulator
apart: both speak exactly this schema. docs/protocol.md documents the framing;
tests/unit/test_protocol.py pins golden bytes that the C# MessageCodec is
validated against (fixture files under tests/fixtures/protocol/).

Channels: sensors | control | capture | truth. The `truth` channel carries
ground truth and may only be subscribed by clients that declared
role="harness" in their hello — this is how the test system guarantees the
localizer never sees the answer.
"""

from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter

PROTOCOL_VERSION = "1.0"

Role = Literal["localizer", "harness", "capture", "verify"]
Channel = Literal["sensors", "control", "capture", "truth"]


class Base(BaseModel):
    model_config = ConfigDict(extra="forbid")


class BlobSpec(Base):
    """Describes one binary blob following the JSON header, in order."""

    name: str
    dtype: Literal["uint8", "float32", "bytes"] = "bytes"
    shape: list[int] | None = None  # None for encoded streams (jpeg/png)
    encoding: Literal["raw", "jpeg", "png"] = "raw"
    byte_len: int


class Header(Base):
    """Fields common to every message."""

    version: str = PROTOCOL_VERSION
    msg_id: int = 0
    sim_time: float = 0.0
    blobs: list[BlobSpec] = Field(default_factory=list)


class Hello(Header):
    type: Literal["hello"] = "hello"
    role: Role = "localizer"


class CameraInfo(Base):
    width: int
    height: int
    fov_deg: float
    tilt_deg: float


class HelloAck(Header):
    type: Literal["hello_ack"] = "hello_ack"
    sim_kind: Literal["unity", "headless"]
    env_id: str
    capabilities: list[str] = Field(default_factory=list)
    geo_meta: dict[str, float] | None = None  # embedded environment geo metadata
    camera: CameraInfo | None = None


class EnvInfoRequest(Header):
    type: Literal["env_info_request"] = "env_info_request"


class EnvInfo(Header):
    type: Literal["env_info"] = "env_info"
    bounds_min_sim: list[float]  # flyable volume, sim frame (x, y, z)
    bounds_max_sim: list[float]
    ground_alt_min_m: float = 0.0
    ground_alt_max_m: float = 0.0


class Subscribe(Header):
    type: Literal["subscribe"] = "subscribe"
    channels: list[Channel]


class SubscribeAck(Header):
    type: Literal["subscribe_ack"] = "subscribe_ack"
    channels: list[Channel]


class CaptureRequest(Header):
    type: Literal["capture_request"] = "capture_request"
    pos_sim: list[float]
    yaw_deg: float  # sim-frame yaw (rotation about sim Y, 0 = +Z)
    pitch_deg: float = 0.0  # camera pitch below horizon (positive = down)
    want: list[Literal["rgb", "depth", "sun"]] = Field(default_factory=lambda: ["rgb"])


class CaptureResult(Header):
    type: Literal["capture_result"] = "capture_result"
    pos_sim: list[float]  # ground-truth pose actually rendered
    quat_sim: list[float]  # (x, y, z, w) body->sim-world
    yaw_deg: float
    pitch_deg: float
    utc: str
    sun_azimuth_deg: float | None = None
    sun_elevation_deg: float | None = None
    moon_azimuth_deg: float | None = None
    moon_elevation_deg: float | None = None
    # blobs: "rgb" (jpeg/png), optional "depth" (float32 HxW)


class SensorFrame(Header):
    type: Literal["sensor_frame"] = "sensor_frame"
    frame_id: int
    utc: str
    lidar_range_m: float | None = None  # None = dropout / out of range
    # blobs: "rgb". Deliberately NO pose fields: the localizer flies blind.


class Command(Header):
    type: Literal["command"] = "command"
    vel_sim: list[float] | None = None  # commanded velocity, sim world frame (m/s)
    yaw_rate_dps: float | None = None


class TruthState(Header):
    type: Literal["truth_state"] = "truth_state"
    frame_id: int
    pos_sim: list[float]
    quat_sim: list[float]
    vel_sim: list[float]
    collided: bool = False


class Reset(Header):
    type: Literal["reset"] = "reset"
    seed: int = 0
    start_pos_sim: list[float] | None = None
    start_yaw_deg: float = 0.0
    utc: str | None = None  # override sim clock start


class ResetDone(Header):
    type: Literal["reset_done"] = "reset_done"
    pos_sim: list[float]


class Teleport(Header):
    """Harness-only: kidnap the drone mid-flight (re-localization test)."""

    type: Literal["teleport"] = "teleport"
    pos_sim: list[float]
    yaw_deg: float = 0.0


class ErrorMsg(Header):
    type: Literal["error"] = "error"
    message: str


class Ping(Header):
    type: Literal["ping"] = "ping"


class Pong(Header):
    type: Literal["pong"] = "pong"


AnyMessage = Annotated[
    Hello | HelloAck | EnvInfoRequest | EnvInfo | Subscribe | SubscribeAck | CaptureRequest | CaptureResult | SensorFrame | Command | TruthState | Reset | ResetDone | Teleport | ErrorMsg | Ping | Pong,
    Field(discriminator="type"),
]

message_adapter: TypeAdapter[AnyMessage] = TypeAdapter(AnyMessage)


def parse_message(header: dict) -> AnyMessage:
    return message_adapter.validate_python(header)
